src/libponyrt/actor/actor.c

```diff
@@ -432,69 +438,124 @@ bool ponyint_actor_run(pony_ctx_t* ctx, pony_actor_t* actor, bool polling)
+      // detector is running whether it is holding a reference to the actor
+      // or not.
+      //
+      // When 'actor_noblock` is true, the cycle detector isn't running.
+      // this means actors won't be garbage collected unless we take special
+      // action. Therefore if `noblock` is on, we should garbage collect the
+      // actor.
+      //
+      // When the cycle detector is running, "deleting locally" is still far
+      // more efficient and will free up memory sooner than letting the cycle
+      // detector do it. However, that is only safe to delete locally if the
+      // cycle detector isn't holding a reference to the actor. If we have never
+      // sent a reference to the cycle detector (FLAG_CD_CONTACTED), then this
+      // is perfectly safe to do.
+      if (actor_noblock || !has_flag(actor, FLAG_CD_CONTACTED))
       {
-        // We're blocked, send block message.
+        // mark the queue as empty or else destroy will hang
+        bool empty = ponyint_messageq_markempty(&actor->q);
+
+        // make sure the queue is actually empty as expected
+        pony_assert(empty);
+
+        // "Locally delete" the actor.
+        ponyint_actor_setpendingdestroy(actor);
+        ponyint_actor_final(ctx, actor);
+        ponyint_actor_destroy(actor);
+
+        // make sure the scheduler will not reschedule this actor
+        return !empty;
+      } else {
+        // The cycle detector is running so we have to ensure that it doesn't
+        // send a message to the actor while we're in the process of telling
+        // it that it is safe to destroy the actor.
+
+        // the cycle detector will not send a message if the actor is marked
+        // pending destroy so mark this actor as pending destroy
+        ponyint_actor_setpendingdestroy(actor);
+
+        // check to make sure the messageq is still empty
+        // and the cycle detector didn't send a message in the meantime
+        if(!ponyint_messageq_isempty(&actor->q))
+        {
+          // need to undo set pending destroy because the cycle detector
+          // sent the actor a message and it is not safe to destroy this
+          // actor as a result.
+          unset_flag(actor, FLAG_PENDINGDESTROY);
+
+          // If we mark the queue as empty, then it is no longer safe to do any
+          // operations on this actor that aren't concurrency safe so make
+          // `ponyint_messageq_markempty` the last thing we do.
+          // Return true (i.e. reschedule immediately) if our queue isn't empty.
+          return !ponyint_messageq_markempty(&actor->q);
+        }
+
+        // At this point the actors queue is empty and the cycle detector
+        // will not send it any more messages because it is set as pending
+        // destroy.
+
+        // the invariant that should hold true at this point is that we have
+        // not sent the cycle detector a `block` message already
+        pony_assert(!has_flag(actor, FLAG_BLOCKED_SENT));
```

Comment:
are you sure? i'm not sure this is correct. i think we might have but don't want to again.

otoh, i really need to think about that, and I'm definitely not sure this isn't correct.

Reply:
so to get here, CD_CONTACTED would have to be set.

And that means we have sent a block at some point. 

There's no guarantee that I see for this that we haven't unblocked and then end up here which is the only way FLAG_BLOCKED_SENT would be set if we get this, right?

In the previous version of this code, this wasn't an assert, it was an if. I think we want an if not an assert here.